### PR TITLE
Add preload tests from Blink.

### PR DESCRIPTION
**Pull request by yoavweiss**
This PR adds tests for [preload](https://w3c.github.io/preload/) from Blink's [layout tests](https://cs.chromium.org/chromium/src/third_party/WebKit/LayoutTests/http/tests/preload/).

These tests are not necessarily a complete test suite as they were built during the feature's implementation, and complemented in Blink by various tests that require internal APIs as well as unit tests. I plan to iterate over these tests and make sure they provide adequate coverage.

/cc @zcorpan @igrigorik

### CODE REVIEW

File: preload/avoid_delaying_onload_link_preload.html [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+    var t = async_test('Makes sure link preload preloaded resources are not delaying onload');
+</script>
+<link rel=preload href="resources/dummy.js?pipe=trickle(d1)" as=script>
```
Comment (zcorpan) [inline preload/avoid_delaying_onload_link_preload.html]: This seems like a possible source of unstable results. Maybe something like 5 seconds? @jgraham

File: preload/single_download_late_used_preload.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+    var t = async_test('Makes sure that preloaded resources are not downloaded again when used');
+</script>
+<div id=result></div>
+<link rel=preload href="resources/square.png" as=image>
+<script>
+    window.addEventListener("load", t.step_func(function() {
+        setTimeout(t.step_func(function() {
```
Comment (zcorpan) [inline preload/single_download_late_used_preload.html]: Use `t.step_timeout` and a longer timeout (2000ms I think)

File: preload/avoid_delaying_onload_link_preload.html (hunk 1 above)
Comment (yoavweiss) [inline preload/avoid_delaying_onload_link_preload.html]: Do you think the time it'd take the browser to run the test and get to window.load() would exceed 1 second? I'd hate to make this test slower if it's not a must

File: preload/delaying_onload_link_preload_after_discovery.html [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+</script>
+<link rel=preload href="resources/dummy.js?pipe=trickle(d1)" as=script>
+<link rel=preload href="resources/square.png?pipe=trickle(d1)" as=image>
+<script>
+    window.addEventListener("load", t.step_func(function() {
+        assert_equals(performance.getEntriesByType("resource").length, 4);
+        t.done();
+    }));
+    document.write('<sc' + 'ript src="resources/dummy.js?pipe=trickle(d1)"></sc' + 'ript>');
+    document.write("<!--");
```
Comment (igrigorik) [inline preload/delaying_onload_link_preload_after_discovery.html]: What's the point/purpose of the unclosed comment? Would it be simpler and cleaner to simply build and append a script element?

File: preload/dynamic_remove_preload_href.html [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+<script>
+    var link = document.createElement("link");
+    link.as = "script";
+    link.rel = "preload";
+    link.href = "resources/dummy.js?pipe=trickle(d0.3)";
+    document.body.appendChild(link);
+    link.href = "";
+    window.addEventListener("load", t.step_func(function() {
+        assert_equals(performance.getEntriesByType("resource").length, 3);
+        t.done();
```
Comment (igrigorik) [inline preload/dynamic_remove_preload_href.html]: Not sure I follow this test. How does this verify that the preload response is aborted? Meta: does abort even make sense? The browser should start a new download if href is mutated, but it's not clear that the old one can/should be stopped.

File: preload/dynamic_removing_preload.html [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    link.rel = "preload";
+    link.href = "resources/dummy.js";
+    document.body.appendChild(link);
+    document.body.removeChild(link);
+    window.addEventListener("load", t.step_func(function() {
+        assert_equals(performance.getEntriesByType("resource").length, 3);
+        t.done();
+    }));
+</script>
+<script src="resources/dummy.js?pipe=trickle(d0.2)"></script>
```
Comment (igrigorik) [inline preload/dynamic_removing_preload.html]: Ditto to above.

File: preload/onerror_event.html [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    var noTypeFailed = false;
+</script>
+<link rel=preload href="non-existent/dummy.js" as=script onerror="scriptFailed = true;">
+<link rel=preload href="non-existent/dummy.css" as=style onerror="styleFailed = true;">
+<link rel=preload href="non-existent/square.png" as=image onerror="imageFailed = true;">
+<link rel=preload href="non-existent/Ahem.ttf" as=font crossorigin onerror="fontFailed = true;">
+<link rel=preload href="non-existent/test.mp4" as=video onerror="videoFailed = true;">
+<link rel=preload href="non-existent/test.oga" as=audio onerror="audioFailed = true;">
+<link rel=preload href="non-existent/security/captions.vtt" as=media onerror="trackFailed = true;">
+<link rel=preload href="non-existent/dummy.xml" as=foobarxmlthing onerror="gibrishFailed = true;">
```
Comment (igrigorik) [inline preload/onerror_event.html]: gibberish? :P

File: preload/delaying_onload_link_preload_after_discovery.html (hunk 3 above)
Comment (yoavweiss) [inline preload/delaying_onload_link_preload_after_discovery.html]: yeah, it would be. I'll switch to that.

File: preload/onerror_event.html (hunk 6 above)
Comment (yoavweiss) [inline preload/onerror_event.html]: didn't bother to typo check that word, apparently :D

File: preload/single_download_preload.html [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    xhr.open("GET", "resources/dummy.xml");
+    xhr.send();
+
+    window.addEventListener("load", t.step_func(function() {
+        // Audio and video show 2 extra requests as the main request is followed by a range request
+        assert_equals(performance.getEntriesByType("resource").length, 14);
+        t.done();
+    }));
+</script>
+<span>PASS - this text is here just so that the browser will download the font.</span
```
Comment (nicjansma) [inline preload/single_download_preload.html]: Missing `>`!

Comment (yoavweiss) [inline preload/single_download_preload.html]: Good catch! :)

**Review by igrigorik — APPROVED** [review]

**Comment by zcorpan** [comment]
https://github.com/w3c/web-platform-tests/pull/4316#discussion_r92421481 is still outstanding comment. (Making tests take longer to run is bad, but having people spend time analyzing unstable results is worse...)

**Comment by yoavweiss** [comment]
Changed the trickle values to higher ones. I'll try to see in follow up PRs if I can somehow collapse the tests, so that we'd wait on these long timeouts once, rather than multiple times...

**Comment by zcorpan** [comment]
Thanks!

Checking test results in https://travis-ci.org/w3c/web-platform-tests/builds/184496213 ...

/preload/preload_with_type.html fails in Chrome. Is that expected?

/preload/dynamic_adding_preload.html times out in Firefox. Could it be made to fail reliably (maybe fail early with a feature-check?)?

**Comment by yoavweiss** [comment]
The failure in Chrome is expected (Noticed it while writing the tests). I can fix it first before landing the tests, if that helps. I'll look into the Firefox timeouts

**Comment by zcorpan** [comment]
It's perfectly fine to land tests that fail, just wanted to check that it was expected.

File: preload/dynamic_adding_preload.html [hunk 8]
```diff
@@ -7,13 +7,17 @@
 <body>
 <script>
     var link = document.createElement("link");
+    if(!link.rellist || !link.rellist.supports("preload")) {
```
Comment (zcorpan) [inline preload/dynamic_adding_preload.html]: Should be `relList`

File: preload/dynamic_adding_preload.html [hunk 9]
```diff
@@ -7,13 +7,17 @@
...
+        assert_unreached("rellist is not defined or browser doesn't support preload.");
+        t.done();
+    }
     link.as = "script";
     link.rel = "preload";
     link.href = "resources/dummy.js";
-    link.onload = t.step_func(function() {
+    link.onload = t.step_timeout(function() {
```
Comment (zcorpan) [inline preload/dynamic_adding_preload.html]: How does this make sense?

File: preload/dynamic_adding_preload.html [hunk 10]
```diff
@@ -7,13 +7,17 @@
 <body>
 <script>
     var link = document.createElement("link");
+    if(!link.rellist || !link.rellist.supports("preload")) {
+        assert_unreached("rellist is not defined or browser doesn't support preload.");
+        t.done();
```
Comment (zcorpan) [inline preload/dynamic_adding_preload.html]: No need for this line, the line above throws an exception.

**Review by zcorpan — CHANGES_REQUESTED** [review]

File: preload/dynamic_adding_preload.html (hunk 9 above)
Comment (yoavweiss) [inline preload/dynamic_adding_preload.html]: There's no spec guaranty that the PerformanceEntry is added to the timeline before the resource's onload handler runs

**Comment by yoavweiss** [comment]
@z

**Comment by yoavweiss** [comment]
@zcorpan: addressed your comments

**Comment by zcorpan** [comment]
w3c-test:mirror

File: preload/dynamic_adding_preload.html [hunk 11]
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script>
+    var t = async_test('Makes sure that a dynamically added preloaded resource is downloaded');
+</script>
+<body>
+<script>
+    var link = document.createElement("link");
```
Comment (yoavweiss) [inline preload/dynamic_adding_preload.html]: done

**Review by zcorpan — APPROVED** [review]

**Comment by yoavweiss** [comment]
Thanks @igrigorik and @zcorpan for reviewing! (and @sideshowbarker for merging :D)

**Pull request merged**